### PR TITLE
[Tests] tally benches

### PR DESCRIPTION
**Pull request by dkijania**
- Generate a number of votes and apply them to the ledger
- Benchmark the creation of the encrypted tally by running it multiple times without saving the result
- Save the result and benchmark the decryption step and applying a decrypted result in the same way

### CODE REVIEW

File: chain-impl-mockchain/benches/tally.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        .vote_encrypted_tally(&alice, encrypted_tally);
+
+    let parameters = ledger.parameters.clone();
+    let date = ledger.date();
+
+    c.bench_function("vote_encrypted_tally", |b| {
+        b.iter(|| {
+            ledger
+                .ledger
+                .apply_fragment(&parameters, &fragment, date)
```
Comment (Zeegomo) [inline chain-impl-mockchain/benches/tally.rs:127]: It's not super clear at first sight why this does not save the result for me, so maybe add a comment?

File: chain-impl-mockchain/benches/tally.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        .owns(STAKE_POOL)
+        .committee_member();
+    wallets.push(&mut alice_wallet_builder);
+
+    let voters_aliases = voters_aliases(VOTERS_COUNT);
+    let mut wallet_builders: Vec<WalletTemplateBuilder> =
+        voters_aliases.iter().map(|alias| wallet(alias)).collect();
+
+    for wallet_builder in wallet_builders.iter_mut() {
+        wallet_builder.with(10);
```
Comment (eugene-babichenko) [inline chain-impl-mockchain/benches/tally.rs:55]: IMO we should increase stakes to be more realistic. I understand that this is going to slow down the tallying process as it is now, but this should not be the case after offline tallying is merged.

**Review by eugene-babichenko — COMMENTED** [review]
I think we should postpone this PR a little bit and rebase it when #250 is merged. The tallying process is going to be changed a bit so does this benchmark. Otherwise, this is a good job.

Also, I think this is the time to introduce the repository-wide documentation on running benchmarks, since it is a bit annoying to remember that you should add `--feature with-bench` every single time.

**Comment by eugene-babichenko** [comment]
The build on the latest nightly compiler is affected by a mistake in `cryptoxide`, I filed a PR to fixed that. https://github.com/typed-io/cryptoxide/pull/15

**Comment by Zeegomo** [comment]
> I think we should postpone this PR a little bit and rebase it when #250 is merged. 

you meant #520?

**Comment by eugene-babichenko** [comment]
Yes.

File: chain-impl-mockchain/benches/tally.rs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    });
+
+    ledger.apply_fragment(&fragment, ledger.date()).unwrap();
+}
+
+fn tally_benchmark_128_voters_1000_ada(c: &mut Criterion) {
+    tally_benchmark(128, 1000, c);
+}
+
+fn tally_benchmark_200_voters_1000_ada(c: &mut Criterion) {
```
Comment (Zeegomo) [inline chain-impl-mockchain/benches/tally.rs:174]: This seems quite similar to the test case above it

**Review by Zeegomo — APPROVED** [review]

File: chain-impl-mockchain/benches/tally.rs (hunk 3 above)
Comment (eugene-babichenko) [reply chain-impl-mockchain/benches/tally.rs:174]: The point was to test with different total voting power. For the previous implementation the voting power made a huge difference.

File: chain-impl-mockchain/benches/tally.rs [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        .iter()
+        .find(|c_vote_plan| {
+            let vote_plan: VotePlan = vote_plan.clone().into();
+            c_vote_plan.id == vote_plan.to_id()
+        })
+        .unwrap();
+
+    c.bench_function(&format!("decrypt_tally{}", bench_name_suffix), |b| {
+        b.iter(|| {
+            decrypt_tally(vote_plan_status, &members);
```
[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply chain-impl-mockchain/benches/tally.rs:147]: So what you suggest is to break down this step into producing shares and then producing decrypted tally out of them? Actually, that's a good point, such benchmark would make much more sense.

File: chain-impl-mockchain/benches/tally.rs [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+    let mut wallet_builders: Vec<WalletTemplateBuilder> =
+        voters_aliases.iter().map(|alias| wallet(alias)).collect();
+
+    for wallet_builder in wallet_builders.iter_mut() {
+        wallet_builder.with(voting_power_per_voter);
+        wallets.push(wallet_builder);
+    }
+
+    let mut rng = ChaCha20Rng::from_seed([0u8; 32]);
+    let members = CommitteeMembersManager::new(&mut rng, THRESHOLD, MEMBERS_NO);
```
[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply chain-impl-mockchain/benches/tally.rs:59]: I do not think that having 1000 members is a realistic scenario to be honest.

[the remark this answers is not in the document]
Comment (Zeegomo) [reply chain-impl-mockchain/benches/tally.rs:59]: Agree, just pointed that out. We can skip this

File: chain-impl-mockchain/benches/tally.rs (hunk 3 above)
Comment (Zeegomo) [reply chain-impl-mockchain/benches/tally.rs:174]: What I mean is that the number of voters does not influence the decryption speed. The advantage of having two benchmarks with similar values in both voters and total voting power (`128_voters_1000_ada` and `200_voters_1000_ada`) is not much IMHO if we already have  bigger tests like `200_votes_1_000_000_ada` or `1000_voters_1000_ada`, but maybe I am missing something. 

Anyway, this is just a nit.

File: chain-impl-mockchain/benches/tally.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Zeegomo) [reply chain-impl-mockchain/benches/tally.rs:147]: Yep, if we want to have an accurate benchmark, that would be a better representation of the actual workload

File: chain-impl-mockchain/benches/tally.rs (hunk 3 above)
Comment (eugene-babichenko) [reply chain-impl-mockchain/benches/tally.rs:174]: > What I mean is that the number of voters does not influence the decryption speed

But it does influence the speed of construction of the encrypted tally.

**Review by Zeegomo — APPROVED** [review]
Great work!

**Pull request merged**